pkg/streamingpromql/operators/selectors/range_vector_selector.go

```diff
@@ -83,35 +104,128 @@ func (m *RangeVectorSelector) NextStepSamples(ctx context.Context) (*types.Range
+	originalRangeEnd := rangeEnd
+
+	// When the smoothed/anchored modifiers are used, the selector (fillBuffer) will return a wider range of points.
+	// Modify the range boundaries accordingly so that we do not discard these extended points.
+	if m.anchored {
+		rangeStart -= m.Selector.LookbackDelta.Milliseconds()
+	} else if m.smoothed {
+		rangeStart -= m.Selector.LookbackDelta.Milliseconds()
+		rangeEnd += m.Selector.LookbackDelta.Milliseconds()
+	}
+
 	m.floats.DiscardPointsAtOrBefore(rangeStart)
 	m.histograms.DiscardPointsAtOrBefore(rangeStart)
+	m.stepData.SmoothedBasisForHeadPointSet = false
+	m.stepData.SmoothedBasisForTailPointSet = false
 
-	if err := m.fillBuffer(m.floats, m.histograms, rangeStart, rangeEnd); err != nil {
+	// When smoothed/anchored is used, the buffer is filled for the user requested time range,
+	// and it may include points immediately outside the range boundaries if they are within the extended time windows.
+	histogramObserved, err := m.fillBuffer(m.floats, m.histograms, originalRangeStart, originalRangeEnd, rangeStart)
+	if err != nil {
 		return nil, err
 	}
 
-	m.stepData.Floats = m.floats.ViewUntilSearchingBackwards(rangeEnd, m.stepData.Floats)
+	if m.anchored || m.smoothed {
+		m.extendedRangeFloats.Release()
+
+		// Histograms are not supported for these modified range queries
+		if histogramObserved {
+			return nil, errors.New("smoothed and anchored modifiers do not work with native histograms")
+		}
+
+		// Note that this view includes the extended range window, so will have the extended points used in the anchored/smoothed filling
+		m.extendedRangeView = m.floats.ViewUntilSearchingBackwards(rangeEnd, m.extendedRangeView)
+
+		// A temporary buffer to capture the extended points.
+		// We do not modify the main ring buffer.
+		// A future optimisation may be to support synthetic points being inserted into the ring buffer, and there by avoid the need for this separate buffer.
+		var buff []promql.FPoint
+		lastInView, hasLastPoint := m.extendedRangeView.Last()
+
+		// If we only find points prior to the start of the original range then no points are returned.
+		if hasLastPoint && lastInView.T > originalRangeStart {
+			if m.smoothed {
+				smoothedPoints, err := NewExtendedPointsForSmoothed(m.extendedRangeView, originalRangeStart, originalRangeEnd, m.memoryConsumptionTracker)
+				if err != nil {
+					return nil, err
+				}
+				m.stepData.SmoothedBasisForHeadPointSet = smoothedPoints.smoothedHeadSet
+				m.stepData.SmoothedBasisForTailPointSet = smoothedPoints.smoothedTailSet
+				if smoothedPoints.smoothedHeadSet {
+					m.stepData.SmoothedBasisForHeadPoint.T = smoothedPoints.smoothedHead.T
+					m.stepData.SmoothedBasisForHeadPoint.F = smoothedPoints.smoothedHead.F
+				}
+				if smoothedPoints.smoothedTailSet {
+					m.stepData.SmoothedBasisForTailPoint.T = smoothedPoints.smoothedTail.T
+					m.stepData.SmoothedBasisForTailPoint.F = smoothedPoints.smoothedTail.F
+				}
```

Comment:
There's no need to assign each field individually:

```suggestion
				if smoothedPoints.smoothedHeadSet {
					m.stepData.SmoothedBasisForHeadPoint = smoothedPoints.smoothedHead
				}
				if smoothedPoints.smoothedTailSet {
					m.stepData.SmoothedBasisForTailPoint = smoothedPoints.smoothedTail
				}
```